### ISSUE TITLE
No contact name

### ISSUE THREAD

**Issue by sphh** [Sanmeet007/logger#16]
I just installed your phone logger from F-Droid.

It does not show the contact name in any of the phone list. It also never asked me for permission to access the contacts.

My contacts are all synchronized with DAVx⁵.

I use the (rather old) Android 8.1.

What do I have to do, that the phone logger displays the contact name instead of just the phone number?

**Comment by Sanmeet007** [comment]
Hi @sphh,

Thank you for your feedback and for trying out the Phone Logger app!

To clarify, the app doesn't require access to your phone's contacts because call logs are stored as a cached database on your device. This approach ensures faster access and display of data while avoiding potential performance issues that could arise from fetching contact details directly.

As for your concern about Android 8.1, the app should work regardless of the Android version. I regularly test it on Android 7, so compatibility with older versions shouldn't be an issue.

**Comment by sphh** [comment]
Thank you for your explanation.

From the `bug` tag I conclude that the contact names should be displayed? Anything I can help with to find the cause of the problem?

**Comment by Sanmeet007** [comment]
@sphh Thank you for your concern! You've already helped a lot just by sharing your feedback.

I marked this as a bug because it’s unexpected behavior for contact names not to appear. Based on my current understanding (though I’m not entirely sure), it seems your contacts might not be saved on your device or synced with your phone, possibly due to privacy settings.

You can try saving one of the contacts directly in your phone’s official Google Contacts app. If the contact name then appears in the app, it’s likely a sync issue unrelated to Logger (and something I can’t fix). Let me know how it goes!

**Comment by sphh** [comment]
My [contact app](https://github.com/FossifyOrg/Contacts) does display the contacts. So does the history in the dialers, I have used: (Fossify Phone)[https://github.com/FossifyOrg/Phone] and [Right Dialer](https://github.com/Goodwy/Dialer/). And IRC the standard phone app as well (when I still used it).

Sorry, I don't want to use the Google Contacts app because of known reasons, so I cannot help you in this department.

**Comment by Sanmeet007** [comment]
@sphh That’s totally fine if you prefer not to use the Google Contacts app. I’ll try to recreate your phone’s environment and work on reproducing the issue. Once I figure it out, I’ll do my best to resolve it. Apologies for the wait, and thank you for your patience!

**Comment by Sanmeet007** [comment]
Hi @sphh,  

Thanks for your feedback! I tested the apps you mentioned, except for DAVx⁵ (I’m not familiar with that one), and everything seems to work as expected on my end. The contact names are being displayed without issues.  

Could you provide more specific steps to reproduce the problem? It might help narrow down the issue. Also, please try relaunching the app—there’s a chance your contacts weren’t synced properly the first time.  

Looking forward to hearing back so we can resolve this for you! 😊

**Comment by Sanmeet007** [comment]
Hi @sphh,  

I tested DAVx⁵ with my Google account and couldn’t reproduce the issue with contact names not displaying. Based on my findings, this seems to be a sync issue with DAVx⁵ rather than a problem with the Logger app.  

I’ll be closing this issue for now, but feel free to comment if you discover anything else.

**Comment by sphh** [comment]
Thanks for investigating.

I now de-installed the app, installed it again from f-droid. No names. I then removed it again and installed it from github. Still no names.

Is there a way to get the database with the call history onto my desktop compute with the help of `adb`, so that I can investigate, if it contains the names or just the numbers?

**Comment by Sanmeet007** [comment]
Okay here is the command 

```
adb shell content query --uri content://call_log/calls
```

In case you are on linux use 
```
 mkdir ~/my_call_logs
 cd ~/my_call_logs
 adb shell content query --uri content://call_log/calls > logs.txt
```

**Comment by sphh** [comment]
Ok. But I get the following permission error:
```
$ adb shell content query --uri content://call_log/calls
Error while accessing provider:call_log
java.lang.SecurityException: Permission Denial: opening provider com.android.providers.contacts.CallLogProvider from (null) (pid=23268, uid=2000) requires android.permission.READ_CALL_LOG or android.permission.WRITE_CALL_LOG
	at android.os.Parcel.readException(Parcel.java:2013)
	at android.os.Parcel.readException(Parcel.java:1959)
	at android.app.IActivityManager$Stub$Proxy.getContentProviderExternal(IActivityManager.java:7161)
	at com.android.commands.content.Content$Command.execute(Content.java:441)
	at com.android.commands.content.Content.main(Content.java:664)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:285)
```

**Comment by Sanmeet007** [comment]
@sphh  In case you succeed try this code in python to convert the file into csv which can be viewed on your excel / libre / google sheets

```py
import pandas as pd, re

TARGET_FILE_PATH = "file.txt"
PARSED_FILE_PATH = "parsed.csv"


def map_to_dict(log_string: str):
    pattern = r"(\w+)=([^,]+)"
    matches = re.findall(pattern, log_string)
    parsed_data = {key: value.strip() for key, value in matches}
    d = {}
    for key, value in parsed_data.items():
        d[key] = value

    return d


li = []

with open(TARGET_FILE_PATH) as file:
    li = list(map(map_to_dict, file.readlines()))

df = pd.DataFrame.from_dict(li)

# Print the columns
print(df.head().columns)

df.to_csv(PARSED_FILE_PATH)

```

**Comment by Sanmeet007** [comment]
@sphh  Just saw your error . Let me see any solutions ... so far i got this only

https://stackoverflow.com/questions/59434725/retrieve-call-logs-android-throught-adb

**Comment by Sanmeet007** [comment]
Ok maybe a solution : 

The error occurs because the `adb shell content query` command requires either the `READ_CALL_LOG` or `WRITE_CALL_LOG` permissions to access the call log data. These permissions are sensitive and protected in Android for security reasons.

To fix this, you need to grant the necessary permissions to the app or process trying to access the call log. Here's how you can resolve it:

### **1. Use `adb` to Grant the Permission**
You can grant the required permission to the shell user (`adb`) using the following command:

```bash
adb shell pm grant com.android.shell android.permission.READ_CALL_LOG
```

Alternatively, if you're trying to write to the call log, grant the `WRITE_CALL_LOG` permission:

```bash
adb shell pm grant com.android.shell android.permission.WRITE_CALL_LOG
```

### **2. Verify the Permission**
After granting the permission, you can verify it with the following command:

```bash
adb shell pm list permissions -g | grep CALL_LOG
```

### **3. Ensure You're Using the Correct `adb` Version**
If you're still experiencing issues, ensure your `adb` version is up-to-date. Update it with the following steps:
- Download the latest version of the Android SDK Platform Tools from [[Google's official site](https://developer.android.com/studio/releases/platform-tools)](https://developer.android.com/studio/releases/platform-tools).
- Replace your existing `adb` with the updated version.

### **4. Check Android Version Restrictions**
Some Android versions (starting from Android 10) require that permissions like `READ_CALL_LOG` can only be granted to **system-level apps** or apps signed with the same key as the system. If this applies to your case:
- You’ll need root access to bypass these restrictions.
- Alternatively, use a custom ROM that allows these permissions to be managed.

### **5. Run as Root (Optional)**
If the above doesn’t work and your device is rooted, you can execute the command with root privileges:

```bash
adb root
adb shell content query --uri content://call_log/calls
```

**Comment by sphh** [comment]
Ad 1. No luck :disappointed: 
```
$ adb shell pm grant com.android.shell android.permission.READ_CALL_LOG
Operation not allowed: java.lang.SecurityException: Package com.android.shell has not requested permission android.permission.READ_CALL_LOG
```

Ad 2.
```
[...]

group:android.permission-group.PHONE
  permission:android.permission.READ_CALL_LOG
  [...]
  permission:android.permission.WRITE_CALL_LOG

[...]

ungrouped:
  [...]
  permission:android.permission.SEND_CALL_LOG_CHANGE

[...]
```

Ad 3.
```
$ adb version
Android Debug Bridge version 1.0.41
Version 34.0.4-debian
Installed as /usr/lib/android-sdk/platform-tools/adb
[...]
```

Ad 4. I have the rather outdated Android 8.1 on my phone.

Ad 5. I never managed to root my phone, hence
```
$ adb root
adbd cannot run as root in production builds
```

**Comment by Sanmeet007** [comment]
@sphh I don't think its android version issue and root out of the box let me check any other possible solutions to extract the call log database ( sorry for spoiling your day )

**Comment by Sanmeet007** [comment]
@sphh How about trying another app that extracts contact information for verification? Additionally, my app allows you to download logs (parsed directly from the database). If the `name`  field in the logs is empty or null, it could point to an issue with DAVx⁵. Let me know if this helps!

**Comment by sphh** [comment]
Thank you for your patience.

I already exported the call log to a csv file and checked it. The `name` field is always `null`.

I also exported the call log with the help of [Right Dialer](https://github.com/Goodwy/Dialer/). There the name field is filled with the proper contact name.

BTW, another idea: Which program writes the call log database? Android itself or the dialer used to initiate/pick up the call? If it is the dialer, [Right Dialer](https://github.com/Goodwy/Dialer/) might not write that information into the database? :thinking: We will know, as soon as I gain access to the db …

**Comment by Sanmeet007** [comment]
@sphh Thank you for your detailed feedback! It seems like we’re both diving deep into solving this issue. I’ll take a closer look at the code for the Right Dialer app, as you suggested, to better understand its implementation. If I find anything relevant, I’ll update you. Let’s get to the bottom of this!

**Comment by Sanmeet007** [comment]
@sphh  From this code from right dialer ( yupp i read the app source code ) 


```kotlin
private fun callContact(useSimOne: Boolean) {
    val phoneNumber = getSelectedPhoneNumber() ?: return
    val name = getSelectedName() ?: return

    activity.callContactWithSimWithConfirmationCheck(phoneNumber, name, useSimOne)
}

private fun callContact(prefix: String = "") {
    val phoneNumber = getSelectedPhoneNumber() ?: return
    val name = getSelectedName() ?: return

    (activity as SimpleActivity).startCallWithConfirmationCheck("$prefix$phoneNumber", name)
}
```
link to the [file](https://github.com/Goodwy/Dialer/blob/ce46259b88d941e0bad97d3b1cd732046a347117/app/src/main/kotlin/com/goodwy/dialer/adapters/RecentCallsAdapter.kt#L4) and line number is 225 



I can concretely confirm that the contact name is being fetched from the **Contacts database** (via the appropriate permissions), not from the call history database. The call history is only used to retrieve the phone number for quicker access to recent calls.

On the other hand, Right Dialer fetches call history and then separately queries the Contacts database to match the contact name with the phone number. While this approach works, it can be much more performance-dependent, especially if you're dealing with a large number of contacts.

For my app, this approach would significantly impact performance and ease of use. Therefore, I’m using the call history for fast access, while contact names are retrieved only when necessary.

So, in conclusion, I believe that the issue you're facing with missing contact names (i.e., null values) is likely due to Davx5 not syncing the contact names properly in your database.

**Comment by Sanmeet007** [comment]
@sphh For further discussion, feel free to email me at ssanmeet123@gmail.com.

**Comment by sphh** [comment]
Thanks for investigating!

Well, that makes sure, that the contact details are displayed. My launcher [KISS](https://github.com/Neamar/KISS) also displays the proper contact name.

I also tried to show the history with the stock Android dialer (`Phone`) from Google. This displays the contact name instantaneously (and not after a long delay like `Right Dialer`). Is this an indication, that `Phone` takes the contact name from the call log?

I don't understand, why DAVx⁵'s synchronization is the culprit, but I am not an expert at all and do not know, how the contact names make their way into the call log. Is the contact name saved by the dialer or by Android itself?

**Comment by Sanmeet007** [comment]
@sphh hank you for your detailed explanation and testing with different apps. Based on your observations, it seems the contact name is coming from the system's call log and synced contacts. The delay you experienced in Right Dialer may be due to how it fetches and matches contact names, which can be slower compared to the stock dialer app.

Regarding DAVx⁵, the issue is likely related to how the app synchronizes contacts, which can cause delays or inconsistencies in displaying contact names in the call log. However, the contact name display is generally handled by the Android system and the dialer app itself.

The call logs themselves often contain cached names, which may not always be up-to-date and are refreshed periodically. After updating a contact name, another sync may be needed to reflect the changes !

As for the issue, I believe it's unrelated to my app. While I could implement an update that requires contact access to read each number and contact name and display the correct contacts in the logs, doing so would severely affect performance, which is against the purpose of my app. My app is designed to efficiently display and manage call logs, and adding this extra processing would compromise that goal.

**Comment by sphh** [comment]
I do understand, why you don't want to fetch the contact name from the contacts. That delay in `Right Dialer` is something I am not too happy with.

Until I can download and inspect the call log, we will not know, if the logs contain the contact names or not. Do I see that correctly?

Do you have any further ideas, why `adb` is refused to access the logs?

**Comment by Sanmeet007** [comment]
@sphh,

Here are a few important facts to consider:

- Your Android device is responsible for syncing the contact name as CACHED_NAME in the call logs.
- Your device syncs call logs and associated contacts periodically.
- If the CACHED_NAME field is null, it likely means your Android device couldn't find the associated contact or is waiting for the next sync.

What the heck is this CACHED_NAME read here : https://developer.android.com/reference/android/provider/CallLog.Calls#CACHED_NAME

So, even after inspecting the call logs, you might find that the contact name is null. In this case, arguing that the contact names are showing as unknown is understandable.

Regarding the ADB access issue, it's normal for Android to restrict access to call logs due to privacy concerns. This is a sensitive area of data, and Android takes measures to protect the call logs database. You can try accessing the contacts database for more insight into how contacts are handled. 

I am no expert at DAVx5 but i do know something that it is responsible for syncing the contacts ! ( so that is the reason i think its culprit )

**Comment by sphh** [comment]
I just started a discussion at DAVx⁵ regarding this issue: https://github.com/bitfireAT/davx5-ose/discussions/1196

**Comment by Sanmeet007** [comment]
@sphh  Thank you for opening the issue at DAVx⁵, which I thought might be responsible for the contact name not syncing in call logs. I appreciate your efforts in looking into this.... Lets get to the root !

**Comment by Sanmeet007** [comment]
@sphh From the discussion at DAVx⁵ (thanks to their maintainer), I was finally able to reproduce the "Unknown" field error. It turns out the issue lies with your dialer app, Right Dialer, which doesn’t register the cached_name field!

Here’s the relevant discussion for reference:
https://github.com/bitfireAT/davx5-ose/discussions/1196#discussioncomment-11669227

Let me know if you need further clarification!

**Comment by Sanmeet007** [comment]
@sphh I sincerely apologize for not reaching this conclusion sooner. Even though you mentioned the `Right Dialer`, I didn’t realize that calls made with it don’t register the `cached_name` field, which results in "Unknown" names. While testing, I used the Google Dialer, which automatically fills the `cached_name` field, so I didn’t suspect this issue. Thanks to the DAVx5 team, I was finally able to reproduce the error!

**Comment by sphh** [comment]
Glad, that the cause has been found. And my inkling has been correct (I would have preferred it otherwise!). I'll reach out to the good people of the `Right Dialer` asking them to include the cached name.

Maybe you want to mention this in the README?

**Comment by Sanmeet007** [comment]
I already did! But shhh… it’s part of the new update I’m releasing soon. Oops, I spilled the beans anyway! 😄 Trust me, it’s going to be awesome—I squashed a ton of bugs too!

**Comment by sphh** [comment]
Here is the link to the issue for `Right Dialer`: https://github.com/Goodwy/Dialer/issues/67
